live-build/misc/migration-scripts/dx_apply

```diff
@@ -16,9 +16,190 @@
+#
+TMPDIR=$(mktemp -d -p "/tmp" -t delphix.XXXXXXX)
+FSNAME=$(basename "$TMPDIR")
+zfs create -o canmount=off -o mountpoint=none "$RPOOL/ROOT"
+zfs create -o canmount=noauto -o mountpoint=/ "$RPOOL/ROOT/$FSNAME"
+
+#
+# Temporarily set the mountpoint to point to a temporary
+# root directory, and unpack our cpio archive to it.
+#
+TMP_ROOT="$TMPDIR/root"
+zfs set mountpoint="$TMP_ROOT" "$RPOOL/ROOT/$FSNAME"
+zfs mount "$RPOOL/ROOT/$FSNAME"
+(
+	cd "$TMP_ROOT"
+	cpio -imu
+) <"$ARCHIVE_DIR/os-root.cpio" 2>&1 ||
+	die "failed to unpack os-root.cpio"
+
+#
+# WIP: Add here any operations that generate files from illumos,
+#	that we may want to carry over and use in Linux. Examples
+#	include networking interface configuration.
+#
+
+#
+# Set things up for the FreeBSD bootloader to provide an option for
+# booting into the Linux environment This happens in two steps:
+# [1] We place vmlinuz and initrd from our migration image into
+#     the /boot directory of the current system, so the bootloader
+#     can load them.
+# [2] Add a new option in the bootloader's menu for Linux. This is
+#     done by appending the configuration of the new option in our
+#     current menu file.
+#
+VMLINUZ=$(find "$TMP_ROOT/boot/vmlinuz-"* | sort -r | head -n 1)
+INITRD_IMG=$(find "$TMP_ROOT/boot/initrd.img-"* | sort -r | head -n 1)
+cp "$VMLINUZ" /boot
+cp "$INITRD_IMG" /boot
+
+#
+# This command loads the Linux Kernel Compressed Executable in memory
+# from the OK prompt of the FreeBSD bootloader.
+#
+# - We specify the root filesystem type to be ZFS and its root dataset
+#   the one that we just created for the migration image.
+#
+# - We set the console parameter during load twice. Once for each
+#   console technology:
+#      tty0 - Makes kernel messages appear in the first virtual
+#             terminal.
+#      ttyS0 - Makes kernel messages appear in the frst serial
+#             port. The 115200n8 part means the serial connection
+#             is made at 115200 baud 8n1.
+#
+# - We specify zfsforce=1 because otherwise we would be failing at
+#   pool import because of ZFS's hostname check.
+#
+LOAD_VMLINUZ_OK_ARG=(
+	"/boot/$(basename "$VMLINUZ")"
+	"root=ZFS=$RPOOL/ROOT/$FSNAME"
+	'console=tty0 console=ttyS0,115200n8'
+	'zfsforce=1'
+)
+LOAD_VMLINUZ_OK_CMD="load ${LOAD_VMLINUZ_OK_ARG[*]}"
+
+#
+# This command loads the initial RAM disk as the initial root filessytem
+# which is needed as part of the booting process.
+#
+LOAD_INITRD_OK_CMD="load -t rootfs /boot/$(basename "$INITRD_IMG")"
+
+#
+# This is the command to be issued whenever our new option in the
+# bootloader menu is chosen. It is FICL code (a Forth-inspired
+# command language) that is basically read as:
+# 	arg func; arg func; func.
+# In our case we:
+# [1] Evaluate the command to load vmlinuz
+# [2] Evaluate the command to load initrd
+# [3] boot (with the above two loaded)
+#
+MAIN_MENU_FICL=(
+	's\"' $LOAD_VMLINUZ_OK_CMD'\"' evaluate
```

Comment:
I don't know much about `forth`, so it's more of a question:
Why do we need the 's\"' stuff and the evaluate. cannot we just pass:
```
<arg-to-load> load <arg-to-2nd-load> load boot
```
Also don't we need any semi-colons between the commands?

Looking at the output you've posted in the PR description, it seems like the escaping sequences for those 2 commands are different:
```
set mainmenu_command[8]="s\" load /boot/vmlinuz-4.15.0-42-generic root=ZFS=rpool/ROOT/delphix.C3zxml7 console=tty0 console=ttyS0,115200n8 zfsforce=1\" evaluate s\" load -t rootfs /boot/initrd.img-4.15.0-42-generic\" evaluate boot"
set menu_timeout_command="s" load /boot/vmlinuz-4.15.0-42-generic root=ZFS=rpool/ROOT/delphix.vI9uLxo console=tty0 console=ttyS0,115200n8 zfsforce=1" evaluate s" load -t rootfs /boot/initrd.img-4.15.0-42-generic" evaluate boot"
```
One of them has `"s\"`, the other one has `"s"`

Reply:
As mentioned in our conversation, these are Forth strings (https://www.complang.tuwien.ac.at/forth/gforth/Docs-html/Characters-and-Strings-Tutorial.html) evaluated within forth.

As for the output line without the slash that was a presentation/editing fault on my part.